ENIGMAsystem/SHELL/Universal_System/Extensions/XInput/XIgamepad.cpp

```diff
@@ -17,238 +16,321 @@
-    XINPUT_VIBRATION vibration;
 
-    // Zeroise the Vibration
-    ZeroMemory(&vibration, sizeof(XINPUT_VIBRATION));
+void gamepad_set_vibration(int device, float left, float right) {
+  if (!gamepad_is_connected(device)) return;
+  XINPUT_VIBRATION vibration = {};
 
-    // Set the Vibration Values
-    vibration.wLeftMotorSpeed = left * 65535;
-    vibration.wRightMotorSpeed = right * 65535;
+  // Set the Vibration Values
+  vibration.wLeftMotorSpeed = left * 65535;
+  vibration.wRightMotorSpeed = right * 65535;
 
-    // Vibrate the controller
-    XInputSetState(device, &vibration);
+  // Vibrate the controller
+  XInputSetState(device, &vibration);
 }
-	
+
 int gamepad_axis_count(int device) {
-	return 2;
+  if (!gamepad_is_connected(device)) return 0;
+  const auto& caps = gamepads[device].caps;
+
+  int axes = 0;
+  if (caps.Gamepad.sThumbLX)
+    ++axes;
+  if (caps.Gamepad.sThumbRX)
+    ++axes;
+  if (caps.Gamepad.sThumbLY)
+    ++axes;
+  if (caps.Gamepad.sThumbRY)
+    ++axes;
+  return axes;
 }
-	
+
 float gamepad_axis_value(int device, int axis) {
-	XINPUT_STATE controllerState;
-    // Zeroise the state
-    ZeroMemory(&controllerState, sizeof(XINPUT_STATE));
-
-    // Get the state
-    XInputGetState(device, &controllerState);
-	float ret;
-	switch (axis) {
-		case gp_axislh:
-			ret = controllerState.Gamepad.sThumbLX;
-			break;
-		case gp_axislv:
-			ret = controllerState.Gamepad.sThumbLY;
-			break;
-		case gp_axisrh:
-			ret = controllerState.Gamepad.sThumbRX;
-			break;
-		case gp_axisrv:
-			ret = controllerState.Gamepad.sThumbRY;
-			break;
-		default:
-			return -1;
-	}
-	if (ret > 0) {
-		return ret / 32767;
-	} else {
-		return ret / 32768;
-	}
-}
-
-WORD digitalButtons[20] = { XINPUT_GAMEPAD_A, XINPUT_GAMEPAD_X, XINPUT_GAMEPAD_B, XINPUT_GAMEPAD_Y, XINPUT_GAMEPAD_LEFT_SHOULDER, 
-	XINPUT_GAMEPAD_RIGHT_SHOULDER, 0, 0, XINPUT_GAMEPAD_BACK, XINPUT_GAMEPAD_START, XINPUT_GAMEPAD_LEFT_THUMB, XINPUT_GAMEPAD_RIGHT_THUMB, 
-	XINPUT_GAMEPAD_DPAD_UP, XINPUT_GAMEPAD_DPAD_DOWN, XINPUT_GAMEPAD_DPAD_LEFT, XINPUT_GAMEPAD_DPAD_RIGHT, 0, 0, 0, 0 };
-	
+  if (!gamepad_is_connected(device)) return 0;
+  const auto& state = gamepads[device].state;
+
+  float ret = 0;
+  bool vertical = false;
+  switch (axis) {
+    case gp_axislh:
+      ret = state.Gamepad.sThumbLX;
+      break;
+    case gp_axislv:
+      ret = state.Gamepad.sThumbLY;
+      vertical = true;
+      break;
+    case gp_axisrh:
+      ret = state.Gamepad.sThumbRX;
+      break;
+    case gp_axisrv:
+      ret = state.Gamepad.sThumbRY;
+      vertical = true;
+      break;
+    default:
+      return 0;
+  }
+  if (ret > 0) {
+    ret /= 32767;
+  } else {
+    ret /= 32768;
+  }
+  const float deadzone = gamepads[device].axis_deadzone;
+  if (ret >= deadzone) {
+    ret = (ret - deadzone) / (1.0f - deadzone);
+  } else if (ret <= -deadzone) {
+    ret = (ret + deadzone) / (1.0f - deadzone);
+  } else {
+    return 0;
+  }
+  return (vertical ? -ret : ret);
```

Comment:
Is this seriously the GM behavior? Best not to lie to people about what axis the joystick is reporting. Let the developer decide that the axis bound to up/down should be inverted.